libraries/exoplayer/src/main/java/androidx/media3/exoplayer/audio/AudioCapabilities.java

```diff
@@ -58,8 +58,9 @@ public final class AudioCapabilities {
   @SuppressWarnings("InlinedApi")
   private static final AudioCapabilities EXTERNAL_SURROUND_SOUND_CAPABILITIES =
       new AudioCapabilities(
-          new int[] {
-            AudioFormat.ENCODING_PCM_16BIT, AudioFormat.ENCODING_AC3, AudioFormat.ENCODING_E_AC3
+          new int[]{
+              AudioFormat.ENCODING_PCM_16BIT, AudioFormat.ENCODING_AC3, AudioFormat.ENCODING_E_AC3,
+              AudioFormat.ENCODING_DTS, AudioFormat.ENCODING_DTS_HD
```

Comment:
The `EXTERNAL_SURROUND_SOUND_CAPABILITIES` that the new encodings are added to, is available to Amazon or Xiaomi device. Did the manufacturers confirm that they support those new encodings?

Reply:
@tianyif Yes. Xperi worked with both manufacturers to make sure that they support those encodings. Thanks.

Reply:
Acknowledged, thanks!

Reply:
on what API versions did Amazon and Xiaomi include the new format? The code currently assumes they are on API 17+. Do all Amazon and Xiaomi devices with API 17+ support ENCODING_DTS and ENCODING_DTS_HD? Do we need to add further API checks?

Reply:
@christosts I agree. I think it is better to add API checks. Let me check which API level and get back. Thanks.

Reply:
@christosts It will be pretty hard to do the API checks as some of the Xiaomi devices may not be consistent in terms of DTS support across the same API. We will stream to newer versions of XIaomi TV that should have support for ENCODING_DTS and ENCODING_DTS_HD. Is it ok if we leave it at API 17+? Thanks.

Reply:
This is going to regress on apps that already have DTS or DTS_HD encodings included in DASH/HLS content that are played on Amazon and Xiaomi devices that do not support the encodings. Once these apps upgrade to latest of ExoPlayer, the player will attempt to play DTS or DTS_HD via passthrough, which likely is going to fail during creation of AudioTrack later.

I do not know how many apps may fall in this category, so I can't evaluate the impact. Do you have any data on this?

On the other hand, do you know if the Amazon and Xiaomi devices that support the new audio formats are after API 29? It seems safer to change the logic here and query the platform for direct support of DTS and DTS_HD with AudioTrack.isDirectAudioPlaybackSupported(), similar to how it's done for other devices.

Reply:
@christosts I was trying to let isDirectAudioPlaybackSupported() do it's job as well. The issue was deviceMaySetExternalSurroundSoundGlobalSetting() will return EXTERNAL_SURROUND_SOUND_CAPABILITIES (which originally did not include DTS encodings) for all Xiaomi TVs, before isDirectAudioPlaybackSupported() has a chance to run. The Xiaomi TV that I have does support ENCODING_DTS and ENCODING_DTS_HD. But the playback stopped because EXTERNAL_SURROUND_SOUND_CAPABILITIES did not include DTS encodings.
So this part of the code is to rectify that.

If there is another way to solve this problem, I would be most happy to try it out. 

I think Amazon and Xiaomi Android 8.0 API Level 26 and above should support these encodings. ENCODING_DTS and ENCODING_DTS_HD were defined in API 23 onwards.

Regarding the impact, I can give you more data through email as this involves some business information which is best communicated via email or some private channel. 

Thanks

Reply:
> If there is another way to solve this problem, I would be most happy to try it out.

For API 29+, we can refactor the code so that for DTS and DTS_HD we query the platform, re-using Api29.getDirectPlaybackSupportedEncodings(). This means we need to change the method to take an argument for the queried encodings. At the moment this method is querying the platform with some fixed properties for audio format and attributes. If the same assumptions work for the DTS and DTS_HD, then we can change the method to only accept an int[] as argument

> I think Amazon and Xiaomi Android 8.0 API Level 26 and above should support these encodings. ENCODING_DTS and ENCODING_DTS_HD were defined in API 23 onwards.

Do the TVs advertise support for DTS/DTS_HD over HDMI? If yes, we pick the encodings from `intent.getIntArrayExtra(AudioManager.EXTRA_ENCODINGS)` which is used further below, right?

So basically, the code here will return

`EXTERNAL_SURROUND_SOUND_CAPABILITIES (=> PCM, AC3, EAC3) +`
+ `if API >= 29: use directPlaybackSupported([DTS, DTS_HD])`
+ `else look at intent.getIntArrayExtra(AudioManager.EXTRA_ENCODINGS)`


> Regarding the impact, I can give you more data through email as this involves some business information which is best communicated via email or some private channel.

Please send us an email at dev.exoplayer@gmail.com with subject "Issue #335". As I said we need to evaluate the risk of causing a regression to existing apps on Amazon and Xiaomi devices that don't support DTS and DTS_HD

Reply:
@christosts  API 29+ for getDirectPlaybackSupportedEncodings() is safe for ENCODING_DTS and ENCODING_DTS_HD. Also, yes, TVs will advertise support for DTS/DTS_HD over HDMI. We check all Android based TVs for this during certification of the TV. So I think your idea of 

EXTERNAL_SURROUND_SOUND_CAPABILITIES (=> PCM, AC3, EAC3) +
- if API >= 29: use directPlaybackSupported([DTS, DTS_HD])
- else look at intent.getIntArrayExtra(AudioManager.EXTRA_ENCODINGS)

should work.

Reply:
@christosts Are you referring to something like below? There is still bugs that I need to work through. E.g. The concatenated list of encodings contain duplicates which needs to be removed.

static AudioCapabilities getCapabilities(Context context, @Nullable Intent intent) {
    
    if (Util.SDK_INT >= 23 && Api23.isBluetoothConnected(context)) {
      return DEFAULT_AUDIO_CAPABILITIES;
    }
    int[] encodings = {};
    if (deviceMaySetExternalSurroundSoundGlobalSetting()
        && Global.getInt(context.getContentResolver(), EXTERNAL_SURROUND_SOUND_KEY, 0) == 1) {
      encodings = EXTERNAL_SURROUND_SOUND_CAPABILITIES;
    }
    
    int[] encodings2 = {};
    if (Util.SDK_INT >= 29 && (Util.isTv(context) || Util.isAutomotive(context))) {
      encodings2 = Util.nullSafeIntegerArrayConcatenation(encodings,
          Api29.getDirectPlaybackSupportedEncodings());
    }
    if (intent == null || intent.getIntExtra(AudioManager.EXTRA_AUDIO_PLUG_STATE, 0) == 0) {
      if (encodings2.length == 0) {
        return DEFAULT_AUDIO_CAPABILITIES;
      } else {
        return new AudioCapabilities(encodings2, /* defaultValue= */ DEFAULT_MAX_CHANNEL_COUNT);
      }
    }
    encodings = Util.nullSafeIntegerArrayConcatenation(encodings2,
        intent.getIntArrayExtra(AudioManager.EXTRA_ENCODINGS));
    return new AudioCapabilities(
        encodings,
        intent.getIntExtra(
            AudioManager.EXTRA_MAX_CHANNEL_COUNT, /* defaultValue= */ DEFAULT_MAX_CHANNEL_COUNT));
  }

Reply:
@christosts If we use the above code, EXTERNAL_SURROUND_SOUND_CAPABILITIES can remain unchanged (without ENCODING_DTS and ENCODING_DTS_HD).

Reply:
With the suggested change, you need to figure the duplicates as you said, and also you query the platform twice on devices with API 29+.

I was thinking that we'd only change the if-block for the Amazon-XIaomi case, from
```
if (deviceMaySetExternalSurroundSoundGlobalSetting()
        && Global.getInt(context.getContentResolver(), EXTERNAL_SURROUND_SOUND_KEY, 0) == 1) {
  return EXTERNAL_SURROUND_SOUND_CAPABILITIES;
}
```

to something like

```
if (deviceMaySetExternalSurroundSoundGlobalSetting()
        && Global.getInt(context.getContentResolver(), EXTERNAL_SURROUND_SOUND_KEY, 0) == 1) {
  return getExternalSurroundCapabilities(intent);
}

// pseudo-code
private static getExternalSurroundCapabilities(@Nullable Intent intent) {
int [] supportedEncodings;
if (Util.SDK_INT >= 29) {
  supportedEncodings = Api29.getDirectPlaybackSupportedEncodings(DTS, DTS_HD);
 return EXTERNAL_SURROUND_SOUND_CAPABILITIES + supportedEncodings;
}

if (intent != null && intent.getIntExtra(AudioManager.EXTRA_AUDIO_PLUG_STATE, 0) == 0)) {
 ...
}
```

I had in mind that we can change  Api29.getDirectPlaybackSupportedEncodings() to take an argument for which encodings to query. For the Amazon/Xiaomi branch, we only query for DTS/DTS_HD where as where we query for all surround encodings furtherdow, we can pass the argument `ALL_SURROUND_ENCODINGS_AND_MAX_CHANNELS.keySet()` (the type of the argument can be int[], or Set<Integer>, I don't think it matters too much)

Reply:
@christosts  Please see 2bbd9891c2797932a7ee3b7a8bf59dda21e09d2f. I tried to stick as close as possible to your recommendations above.  ALL_SURROUND_ENCODINGS_AND_MAX_CHANNELS.keySet() was refactored to return from getAllSurroundEncodingsMaybeSupported() as proposed by @tianyif  earlier. Let me know. Thanks.